internal/charmstore/migrationsdump_test.go

```diff
@@ -0,0 +1,1012 @@
+)
+
+func (s *migrationsDumpSuite) TestMigrationFromDump(c *gc.C) {
+	s.createDatabaseAtLastMigrationVersion(c)
+	err := s.runMigrations()
+	c.Assert(err, gc.IsNil)
+
+	p, err := NewPool(s.Session.DB(historicalDBName), nil, &bakery.NewServiceParams{}, ServerParams{})
+	c.Assert(err, gc.IsNil)
+	store := p.Store()
+	p.Close()
+	defer store.Close()
+
+	// TODO: lots more checks for all the properties we expect. This is
+	// just an initial smoke test.
+	entity, err := store.FindBestEntity(charm.MustParseURL("development/precise/promulgated"), nil)
+	c.Assert(err, gc.IsNil)
+	c.Assert(entity.PromulgatedURL.String(), gc.Equals, "cs:precise/promulgated-1")
+}
+
+// runMigrations starts a new server which will cause all migrations
+// to be triggered.
+func (s *migrationsDumpSuite) runMigrations() error {
+	db := s.Session.DB(historicalDBName)
+	apiHandler := func(p *Pool, config ServerParams, _ string) HTTPCloseHandler {
+		return nopCloseHandler{http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {})}
+	}
+	srv, err := NewServer(db, nil, serverParams, map[string]NewAPIHandlerFunc{
+		"version1": apiHandler,
+	})
+	if err == nil {
+		srv.Close()
+	}
+	return err
+
+}
+
+// createDatabaseAtLastMigrationVersion changes the state of the
+// database to be as it is after running the database
+// migrations from all the versions and updates specified in migrationHistory.
+// Depending on command-line flags, it might run these
+// from scratch or restore from from a previously created file.
+func (s *migrationsDumpSuite) createDatabaseAtLastMigrationVersion(c *gc.C) {
+	db := s.Session.DB(historicalDBName)
+	lastVersion := migrationHistory[len(migrationHistory)-1].version
+	if *dumpMigrationHistory {
+		*runMigrationHistory = true
+	}
+	if *runMigrationHistory {
+		err := s.runHistory(db, migrationHistory)
+		c.Assert(err, gc.IsNil)
+		if *dumpMigrationHistory {
+			vcsStatus, err := currentVCSStatus()
+			c.Assert(err, gc.IsNil)
+			err = saveDBToFile(db, vcsStatus, migrationDumpFileName(lastVersion))
+			c.Assert(err, gc.IsNil)
+		}
+		return
+	}
+	vcsStatus, err := restoreDBFromFile(db, migrationDumpFileName(lastVersion))
+	c.Assert(err, gc.IsNil)
+	c.Logf("restored migration from version %s; dumped at %s", lastVersion, vcsStatus)
+}
+
+// migrationDumpFileName returns the name of the file that
+// the migration database snapshot will be saved to.
+func migrationDumpFileName(version string) string {
+	return "migrationdump." + version + ".zip"
+}
+
+// currentVCSStatus returns the git status of the current
+// charmstore source code. This will be saved into the
+// migration dump file so that there is some indication
+// as to when that was created.
+func currentVCSStatus() (string, error) {
+	cmd := exec.Command("git", "describe")
+	cmd.Stderr = os.Stderr
+	data, err := cmd.Output()
+	if err != nil {
+		return "", errgo.Mask(err)
+	}
+	cmd = exec.Command("git", "status", "--porcelain")
```

Comment:
perhaps add a comment here saying what it does, in case I don't know what git status --porcelain does.